include/kvs_async_client.hpp

```diff
@@ -0,0 +1,548 @@
+
+    // bind the two sockets we listen on
+    key_address_puller_.bind(ut_.key_address_bind_address());
+    response_puller_.bind(ut_.response_bind_address());
+
+    pollitems_ = {
+        {static_cast<void*>(key_address_puller_), 0, ZMQ_POLLIN, 0},
+        {static_cast<void*>(response_puller_), 0, ZMQ_POLLIN, 0},
+    };
+
+    // set the request ID to 0
+    rid_ = 0;
+  }
+
+  ~KvsAsyncClient() {}
+
+ public:
+  /**
+   * Issue an async PUT request to the KVS for a certain lattice typed value.
+   */
+  void put_async(const Key& key, const string& payload,
+                 LatticeType lattice_type,
+                 const Address& response_addr = string()) {
+    KeyRequest request;
+    KeyTuple* tuple = prepare_data_request(request, key);
+    request.set_type(RequestType::PUT);
+    tuple->set_lattice_type(lattice_type);
+    tuple->set_payload(payload);
+
+    try_request(request, response_addr);
+  }
+
+  /**
+   * Issue an async GET request to the KVS for a cross causal lattice value.
+   */
+  void get_async(const Key& key, const Address& response_addr = string()) {
+    // we issue GET only when it is not in the pending map
+    if (pending_get_response_map_.find(key) ==
+        pending_get_response_map_.end()) {
+      KeyRequest request;
+      prepare_data_request(request, key);
+      request.set_type(RequestType::GET);
+
+      try_request(request, response_addr);
+    }
+  }
+
+  vector<pair<KeyResponse, Address>> receive_async(ZmqUtilInterface* kZmqUtil) {
+    vector<pair<KeyResponse, Address>> result;
+
+    kZmqUtil->poll(0, &pollitems_);
+    // handle a key address response
+    if (pollitems_[0].revents & ZMQ_POLLIN) {
+      string serialized = kZmqUtil->recv_string(&key_address_puller_);
+      KeyAddressResponse response;
+      response.ParseFromString(serialized);
+      Key key = response.addresses(0).key();
+      if (pending_request_map_.find(key) != pending_request_map_.end()) {
+        if (response.error() == 1) {
+          log_->error(
+              "No servers have joined the cluster yet. Retrying request.");
+          pending_request_map_[key].first = std::chrono::system_clock::now();
+          query_routing_async(key);
+        } else {
+          // populate cache
+          for (const Address& ip : response.addresses(0).ips()) {
+            key_address_cache_[key].insert(ip);
+          }
+          // handle stuff in pending request map
+          for (auto& pair : pending_request_map_[key].second) {
+            try_request(pair.first, pair.second);
+          }
+          // GC the pending request map
+          pending_request_map_.erase(key);
+        }
+      }
+    }
+    // handle a response
+    if (pollitems_[1].revents & ZMQ_POLLIN) {
+      string serialized = kZmqUtil->recv_string(&response_puller_);
+      KeyResponse response;
+      response.ParseFromString(serialized);
+      Key key = response.tuples(0).key();
+      if (response.type() == RequestType::GET) {
+        // GET
```

Comment:
Nit: Unnecessary comments. It's clear from the previous. line what the request type is.